apiserver/facades/controller/sshserver/facade.go

```
@@ -61,10 +74,53 @@ func (f *Facade) WatchControllerConfig() (params.NotifyWatchResult, error) {
 // SSHServerHostKey returns the controller's SSH server host key.
 func (f *Facade) SSHServerHostKey() (params.StringResult, error) {
 	result := params.StringResult{}
-	key, err := f.backend.SSHServerHostKey()
+	key, err := f.systemState.SSHServerHostKey()
 	if err != nil {
 		result.Error = apiservererrors.ServerError(err)
 	}
 	result.Result = key
 	return result, nil
 }
+
+// authorizedKeysPerModel collects the authorized keys given a model uuid.
+func (f *Facade) authorizedKeysPerModel(uuid string) ([]string, error) {
+	model, p, err := f.statePool.GetModel(uuid)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer p.Release()
+	cfg, err := model.Config()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	keys := jujussh.SplitAuthorisedKeys(cfg.AuthorizedKeys())
+	return keys, nil
+}
+
+// PublicKeyAuthentication extracts the models a user has access to, get all the authorized keys and search for a match.
+// If it's found it returns true, in case of errors or no-match returns false.
+// It is expected that a key available on one model allows the user to authenticate to any model,
+// but later one where we can check the destination model, we will restrict access to the specific model
+// the user is trying to SSH into.
+func (f *Facade) PublicKeyAuthentication(userTag names.UserTag, publicKey ssh.PublicKey) (bool, error) {
```

We should be using a bespoke type in the params package to hold the arguments. Perhaps called `PublicKeyAuthenticationArgs`.

Additionally, we should not rely on the ssh.PublicKey serialisation/deserialisation here, it would be better for the public key to go over the wire as a string.

---

also, we could just return a specific error in case of `no match` and have a single return - no need for the extra bool

---

done.